pkg/idstore/store_impl.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package idstore
+
+import (
+	"fmt"
+
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/metrics/disabled"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/idstore"
+	"github.com/hyperledger/fabric/core/ledger/util/couchdb"
+	"github.com/hyperledger/fabric/protos/common"
+	"github.com/pkg/errors"
+	"github.com/trustbloc/fabric-peer-ext/pkg/roles"
+)
+
+var logger = flogging.MustGetLogger("idstore")
+
+const (
+	systemID      = "fabric_system_"
+	inventoryName = "inventory"
+)
+
+//Store contain couchdb instance
+type Store struct {
+	db               *couchdb.CouchDatabase
+	couchMetadataRev string
+}
+
+//OpenIDStore return id store
+func OpenIDStore(path string) idstore.IDStore {
+	couchInstance, err := createCouchInstance()
+	if err != nil {
+		logger.Error("create couchdb instance failed %s", err.Error())
+		return nil
+	}
+
+	inventoryDBName := couchdb.ConstructBlockchainDBName(systemID, inventoryName)
+	if roles.IsCommitter() {
+		return newCommitterStore(couchInstance, inventoryDBName)
+	}
+	s, err := newStore(couchInstance, inventoryDBName)
+	if err != nil {
+		logger.Error(err.Error())
+		return nil
+	}
+	return s
+}
+
+func newStore(couchInstance *couchdb.CouchInstance, dbName string) (idstore.IDStore, error) {
+	db, err := couchdb.NewCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "create new couchdb database called [%s] failed", dbName)
+	}
+
+	dbExists, err := db.ExistsWithRetry()
+	if err != nil {
+		return nil, errors.WithMessagef(err, "check couchdb [%s] exist failed %v", dbName)
+	}
+	if !dbExists {
+		return nil, errors.New(fmt.Sprintf("DB not found: [%s]", dbName))
+	}
+
+	indexExists, err := db.IndexDesignDocExistsWithRetry(inventoryTypeIndexDoc)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "check couchdb [%s] index exist failed", dbName)
+
+	}
+	if !indexExists {
+		return nil, errors.New(fmt.Sprintf("DB index not found: [%s]", db.DBName))
+	}
+
+	s := Store{db, ""}
+	return &s, nil
+}
+
+func newCommitterStore(couchInstance *couchdb.CouchInstance, dbName string) idstore.IDStore {
+	db, err := couchdb.CreateCouchDatabase(couchInstance, dbName)
+	if err != nil {
+		logger.Error("create new couchdb database failed %s", err.Error())
+		return nil
+	}
+
+	err = createIndices(db)
+	if err != nil {
+		logger.Error("create couchdb index failed %s", err.Error())
+		return nil
+	}
+
+	s := Store{db, ""}
+
+	return &s
+}
+
+func createIndices(db *couchdb.CouchDatabase) error {
+	err := db.CreateNewIndexWithRetry(inventoryTypeIndexDef, inventoryTypeIndexDoc)
+	if err != nil {
+		return errors.WithMessage(err, "creation of inventory metadata index failed")
```

Comment:
maybe change to:
return errors.WithMessage**f(err, "creation of inventory metadata index failed for [%s]", db.DBName)**

?

Reply:
done

Reply:
done